README.rst
================================================================================
pyecharts-extras
================================================================================

.. image:: https://api.travis-ci.org/pyecharts/pyecharts-extras.svg
   :target: http://travis-ci.org/pyecharts/pyecharts-extras

.. image:: https://codecov.io/github/pyecharts/pyecharts-extras/coverage.png
   :target: https://codecov.io/github/pyecharts/pyecharts-extras
.. image:: https://badge.fury.io/py/pyecharts-extras.svg
   :target: https://pypi.org/project/pyecharts-extras

.. image:: https://pepy.tech/badge/pyecharts-extras/month
   :target: https://pepy.tech/project/pyecharts-extras/month

.. image:: https://img.shields.io/github/stars/pyecharts/pyecharts-extras.svg?style=social&maxAge=3600&label=Star
    :target: https://github.com/pyecharts/pyecharts-extras/stargazers

.. image:: https://mybinder.org/badge_logo.svg
 :target: https://mybinder.org/v2/gh/pyecharts/pyecharts-extras/master?filepath=examples



Latest version works with pyecharts 1.3.1+.

Please launch the binder to get an interactive jupyter notebook running.

1. Time Series Event
-----------------------

Usage of TimeYAxisDateXAxis:

.. code-block:: python

   import pyecharts.options as opts
   from pyecharts_extras.date_time import TimeYAxisDateXAxis, DateAxisOpts, TimeAxisOpts
   from datetime import datetime
   
   data = [
      datetime(2020,9,1,3,11,11),
      ...
   ]
   
   data2 = [
       datetime(2020,9,1,1,11,11),
       ...
   ]
   all_data = data1 + data2
   all_dates = sorted(list(set([str(d.date()) for d in all_data])))
   chart = TimeYAxisDateXAxis(init_opts=opts.InitOpts(height="1000px"))
   chart.set_global_opts(
            xaxis_opts=DateAxisOpts(axis_data=all_dates),
            yaxis_opts=TimeAxisOpts(),
            tooltip_opts=opts.TooltipOpts(is_show=False))   
   chart.add_data('A', data, symbol_size=10)
   chart.add_data('B', data2, symbol_size=20)
   chart.render_notebook()


.. image:: https://user-images.githubusercontent.com/4280312/92998181-4b4f6e00-f510-11ea-9b0a-963a83f64e04.png


2. ChoroplethMap
----------------

.. image:: https://user-images.githubusercontent.com/4280312/61274388-a0281500-a7a3-11e9-9501-27e8a1659bd0.png



Installation
================================================================================


You can install pyecharts-extras via pip:

.. code-block:: bash

    $ pip install pyecharts-extras


or clone it and install it:

.. code-block:: bash

    $ git clone https://github.com/pyecharts/pyecharts-extras.git
    $ cd pyecharts-extras
    $ python setup.py install


CHANGELOG.rst
Change log
================================================================================

0.0.5 - 12.09.2020
--------------------------------------------------------------------------------

**Fixed**

#. time date scatter plot failed to display different sets of data

0.0.4 - 06.09.2020
--------------------------------------------------------------------------------

**Added**

#. it scatters event over time vs date chart so as to spot patterns in events.

0.0.3 - 15.07.2019
--------------------------------------------------------------------------------

**Updated**

#. it use pyecharts 1.3 to draw ChoroplethMap

0.0.2 - 09.06.2019
--------------------------------------------------------------------------------

**Updated**

#. it use pyecharts 1.x to draw ChoroplethMap

0.0.1 - 17.04.2019
--------------------------------------------------------------------------------

**first release**

#. it draws ChoroplethMap


test.sh
pip freeze
nosetests --with-coverage --cover-package pyecharts_extras --cover-package tests tests   pyecharts_extras


tests/test_date_time.py
import codecs
import os
from datetime import datetime

from pyecharts import options as opts

from pyecharts_extras import TimeYAxisDateXAxis


def test_time_vs_date():
    # show label
    data = [
        datetime(2020, 9, 1, 11, 11, 11),
        datetime(2020, 9, 1, 9, 11, 11),
        datetime(2020, 9, 2, 12, 11, 11),
        datetime(2020, 9, 3, 13, 11, 11),
        datetime(2020, 9, 4, 14, 11, 11),
        datetime(2020, 9, 5, 15, 11, 11),
        datetime(2020, 9, 6, 16, 11, 11),
        datetime(2020, 9, 1, 1, 11, 11),
        datetime(2020, 9, 2, 2, 11, 11),
        datetime(2020, 9, 3, 3, 11, 11),
        datetime(2020, 9, 4, 4, 11, 11),
        datetime(2020, 9, 5, 5, 11, 11),
        datetime(2020, 9, 6, 6, 11, 11),
    ]

    data2 = [
        datetime(2020, 8, 1, 19, 11, 11),
        datetime(2020, 8, 2, 13, 19, 11),
        datetime(2020, 8, 3, 13, 19, 11),
        datetime(2020, 8, 4, 13, 19, 11),
        datetime(2020, 8, 5, 13, 19, 11),
        datetime(2020, 9, 6, 21, 19, 11),
        datetime(2020, 9, 3, 13, 25, 11),
        datetime(2020, 9, 4, 14, 19, 11),
        datetime(2020, 9, 5, 17, 19, 11),
        datetime(2020, 9, 6, 23, 19, 11),
    ]
    c = TimeYAxisDateXAxis(init_opts=opts.InitOpts(height="1000px"))
    c.add_data("A", data, symbol_size=10)
    c.add_data("B", data2, symbol_size=10)
    c.render()
    content = get_default_rendering_file_content()
    assert "2020-08-01" in content
    assert "69071.0" in content
    os.unlink("render.html")


def get_default_rendering_file_content(file_name="render.html"):
    """
    Simply returns the content render.html
    """
    with codecs.open(file_name, "r", "utf-8") as f:
        return f.read()


tests/test_choropleth_map.py
import codecs
import os

from pyecharts import options as opts

from pyecharts_extras import ChoroplethMap


def test_choropleth_map():
    # show label
    value = ["A", "B", "C", "THIS_KEY_IS_NOT_IN_HTML", "A"]
    attr = ["福建", "山东", "北京", "上海", "西藏"]
    legends = [
        {"tag": "A", "label": "test a", "color": "blue"},
        {"tag": "B", "label": "test b", "color": "yellow"},
        {"tag": "C", "label": "test c", "color": "green"},
        {"tag": "THIS_KEY_IS_NOT_IN_HTML", "label": "test d", "color": "red"},
    ]
    map = ChoroplethMap(
        init_opts=opts.InitOpts(width="1200px", height="600px")
    )
    map.set_global_opts(
        title_opts=opts.TitleOpts(title="Choropleth map - 等值区域图示例")
    )
    map.add(
        "",
        data_pair=zip(attr, value),
        choropleth_legend=legends,
        maptype="china",
        is_map_symbol_show=False,
        label_opts=opts.LabelOpts(is_show=False),
    )
    map.render()
    content = get_default_rendering_file_content()
    assert "piecewise" in content
    assert "test a" in content
    assert "2.1" in content
    assert "THIS_KEY_IS_NOT_IN_HTML" not in content
    os.unlink("render.html")


def get_default_rendering_file_content(file_name="render.html"):
    """
    Simply returns the content render.html
    """
    with codecs.open(file_name, "r", "utf-8") as f:
        return f.read()


pyecharts_extras/choropleth_map.py
from pyecharts import options as opts
from pyecharts.charts import Map
from pyecharts.types import Sequence


class ChoroplethMap(Map):
    def add(
        self,
        series_name: str,
        data_pair: Sequence,
        maptype: str = "china",
        *,
        choropleth_legend: Sequence = None,
        **kwds
    ):
        _piece_specs = []
        _piece_colors = []
        _classes = []
        for __index, __label in enumerate(choropleth_legend):
            # associate label and color with the
            # unique internal index
            _classes.append(__label["tag"])
            _piece_specs.append(
                {
                    "min": __index,
                    "max": __index + 0.1,
                    "label": __label["label"],
                }
            )
            _piece_colors.append(__label["color"])

        _old_data_pair = list(data_pair)
        _new_data_pair = []
        for key, value in _old_data_pair:
            # exchange text value with unique internal index
            __index__ = _classes.index(value)
            _new_data_pair.append([key, __index__])

        # dictate the arguments for Map.add()
        custom = opts.VisualMapOpts(
            is_piecewise=True,
            range_text=["Legend"],
            range_color=_piece_colors,
            pieces=_piece_specs,
        )
        self.options.update(visualMap=custom)

        super(ChoroplethMap, self).add(
            series_name, _new_data_pair, maptype, **kwds
        )
        return self


pyecharts_extras/date_time.py
from datetime import timedelta

import pyecharts.options as opts
from pyecharts import types
from pyecharts.charts.chart import RectChart
from pyecharts.commons.utils import JsCode
from pyecharts.globals import ChartType
from pyecharts.options.series_options import (
    LabelOpts,
    MinorSplitLineOpts,
    MinorTickOpts,
    Numeric,
    Optional,
    SplitAreaOpts,
    SplitLineOpts,
    TextStyleOpts,
    Union,
)

ONE_HOUR = 3600
ONE_DAY = 24 * ONE_HOUR
TIME_AXIS_LABEL_FORMATTER = """
function (value){
    return new Date(value * 1000).toISOString().substr(11, 8);
}
"""
DEFAULT_TIME_AXIS_LABEL = opts.LabelOpts(
    formatter=JsCode(TIME_AXIS_LABEL_FORMATTER)
)
DEFAULT_TIME_AXIS_TICK = opts.AxisTickOpts(is_show=True)
DEFAULT_TIME_AXIS_SPLIT_LINE = opts.SplitLineOpts(is_show=True)


class TimeAxisOpts(opts.AxisOpts):
    def __init__(
        self,
        type_: Optional[str] = "value",
        name: Optional[str] = None,
        is_show: bool = True,
        is_scale: bool = False,
        is_inverse: bool = False,
        name_location: str = "end",
        name_gap: Numeric = 15,
        name_rotate: Optional[Numeric] = None,
        interval: Optional[Numeric] = ONE_HOUR,
        grid_index: Numeric = 0,
        position: Optional[str] = None,
        offset: Numeric = 0,
        split_number: Numeric = 5,
        boundary_gap: Union[str, bool, None] = None,
        min_: Union[Numeric, str, None] = 0,
        max_: Union[Numeric, str, None] = ONE_DAY,
        min_interval: Numeric = 0,
        max_interval: Optional[Numeric] = None,
        axisline_opts: Union[opts.AxisLineOpts, dict, None] = None,
        axistick_opts: Union[
            opts.AxisTickOpts, dict, None
        ] = DEFAULT_TIME_AXIS_TICK,
        axislabel_opts: Union[LabelOpts, dict, None] = DEFAULT_TIME_AXIS_LABEL,
        axispointer_opts: Union[opts.AxisPointerOpts, dict, None] = None,
        name_textstyle_opts: Union[TextStyleOpts, dict, None] = None,
        splitarea_opts: Union[SplitAreaOpts, dict, None] = None,
        splitline_opts: Union[
            SplitLineOpts, dict
        ] = DEFAULT_TIME_AXIS_SPLIT_LINE,
        minor_tick_opts: Union[MinorTickOpts, dict, None] = None,
        minor_split_line_opts: Union[MinorSplitLineOpts, dict, None] = None,
    ):
        super().__init__(
            type_=type_,
            name=name,
            is_show=is_show,
            is_scale=is_scale,
            name_location=name_location,
            name_gap=name_gap,
            name_rotate=name_rotate,
            interval=interval,
            name_textstyle_opts=name_textstyle_opts,
            grid_index=grid_index,
            axisline_opts=axisline_opts,
            axistick_opts=axistick_opts,
            axislabel_opts=axislabel_opts,
            axispointer_opts=axispointer_opts,
            is_inverse=is_inverse,
            position=position,
            offset=offset,
            split_number=split_number,
            boundary_gap=boundary_gap,
            min_=min_,
            max_=max_,
            min_interval=min_interval,
            max_interval=max_interval,
            splitline_opts=splitline_opts,
            splitarea_opts=splitarea_opts,
            minor_tick_opts=minor_tick_opts,
            minor_split_line_opts=minor_split_line_opts,
        )


class DateAxisOpts(opts.AxisOpts):
    def __init__(
        self,
        type_: Optional[str] = None,
        name: Optional[str] = None,
        is_show: bool = True,
        is_scale: bool = False,
        is_inverse: bool = False,
        name_location: str = "end",
        name_gap: Numeric = 15,
        name_rotate: Optional[Numeric] = None,
        interval: Optional[Numeric] = None,
        grid_index: Numeric = 0,
        position: Optional[str] = None,
        offset: Numeric = 0,
        split_number: Numeric = 5,
        boundary_gap: Union[str, bool, None] = None,
        min_: Union[Numeric, str, None] = None,
        max_: Union[Numeric, str, None] = None,
        min_interval: Numeric = 0,
        max_interval: Optional[Numeric] = None,
        axisline_opts: Union[opts.AxisLineOpts, dict, None] = None,
        axistick_opts: Union[
            opts.AxisTickOpts, dict, None
        ] = opts.AxisTickOpts(is_show=True),
        axislabel_opts: Union[LabelOpts, dict, None] = None,
        axispointer_opts: Union[opts.AxisPointerOpts, dict, None] = None,
        name_textstyle_opts: Union[TextStyleOpts, dict, None] = None,
        splitarea_opts: Union[SplitAreaOpts, dict, None] = None,
        splitline_opts: Union[SplitLineOpts, dict] = SplitLineOpts(),
        minor_tick_opts: Union[MinorTickOpts, dict, None] = None,
        minor_split_line_opts: Union[
            MinorSplitLineOpts, dict, None
        ] = SplitLineOpts(is_show=True),
        axis_data=None,
    ):
        # date_min = time.mktime(min_.timetuple())
        # date_max = time.mktime(max_.timetuple())
        date_min = min_
        date_max = max_
        super().__init__(
            type_=type_,
            name=name,
            is_show=is_show,
            is_scale=is_scale,
            name_location=name_location,
            name_gap=name_gap,
            name_rotate=name_rotate,
            interval=interval,
            name_textstyle_opts=name_textstyle_opts,
            grid_index=grid_index,
            axisline_opts=axisline_opts,
            axistick_opts=axistick_opts,
            axislabel_opts=axislabel_opts,
            axispointer_opts=axispointer_opts,
            is_inverse=is_inverse,
            position=position,
            offset=offset,
            split_number=split_number,
            boundary_gap=boundary_gap,
            min_=date_min,
            max_=date_max,
            min_interval=min_interval,
            max_interval=max_interval,
            splitline_opts=splitline_opts,
            splitarea_opts=splitarea_opts,
            minor_tick_opts=minor_tick_opts,
            minor_split_line_opts=minor_split_line_opts,
        )
        self.opts["data"] = axis_data


class TimeYAxisDateXAxis(RectChart):
    def __init__(self, init_opts: types.Init = opts.InitOpts()):
        super().__init__(init_opts=init_opts)

    def add_data(
        self,
        series_name: str,
        y_axis: list,
        *,
        is_selected: bool = True,
        xaxis_index: types.Optional[types.Numeric] = None,
        yaxis_index: types.Optional[types.Numeric] = None,
        color: types.Optional[str] = None,
        symbol: types.Optional[str] = None,
        symbol_size: types.Union[types.Numeric, types.Sequence] = 10,
        symbol_rotate: types.Optional[types.Numeric] = None,
        label_opts: types.Label = opts.LabelOpts(is_show=False),
        markpoint_opts: types.MarkPoint = None,
        markline_opts: types.MarkLine = None,
        markarea_opts: types.MarkArea = None,
        tooltip_opts: types.Tooltip = None,
        itemstyle_opts: types.ItemStyle = None,
        encode: types.Union[types.JSFunc, dict, None] = None,
    ):
        self._append_color(color)
        self._append_legend(series_name, is_selected)
        pair = [[d.date(), d.time()] for d in y_axis]
        data = [
            (
                str(t[0]),
                timedelta(
                    hours=t[1].hour, minutes=t[1].minute, seconds=t[1].second
                ).total_seconds(),
            )
            for t in pair
        ]

        self.options.get("series").append(
            {
                "type": ChartType.SCATTER,
                "name": series_name,
                "xAxisIndex": xaxis_index,
                "yAxisIndex": yaxis_index,
                "symbol": symbol,
                "symbolSize": symbol_size,
                "symbolRotate": symbol_rotate,
                "data": data,
                "label": label_opts,
                "markPoint": markpoint_opts,
                "markLine": markline_opts,
                "markArea": markarea_opts,
                "tooltip": tooltip_opts,
                "itemStyle": itemstyle_opts,
                "encode": encode,
            }
        )
        return self
